File: bindings/nostr-sdk-js/examples/bot.js
const { Keys, Client, NostrSigner, Filter, UnwrappedGift, initLogger, LogLevel, loadWasmAsync, EventBuilder } = require("../");

async function main() {
    await loadWasmAsync();

    try {
        initLogger(LogLevel.info());
    } catch (error) {
        console.log(error);
    }


    let keys = Keys.parse("nsec1ufnus6pju578ste3v90xd5m2decpuzpql2295m3sknqcjzyys9ls0qlc85");

    let signer = NostrSigner.keys(keys);
    let client = new Client(signer);

    await client.addRelay("wss://relay.damus.io");
    await client.addRelay("wss://nos.lol");
    await client.addRelay("wss://nostr.oxtr.dev");

    await client.connect();

    const filter = new Filter().pubkey(keys.publicKey).kind(1059).limit(0); // Limit set to 0 to get only new events! Timestamp.now() CAN'T be used for gift wrap since the timestamps are tweaked!
    console.log('filter', filter.asJson());

    await client.subscribe([filter]);

    const handle = {
        // Handle event
        handleEvent: async (relayUrl, subscriptionId, event) => {
            console.log("Received new event from ", relayUrl);
            if (event.kind === 1059) {
                try {
                    let content = await UnwrappedGift.fromGiftWrap(signer, event);
                    let sender = content.sender;
                    let rumor = content.rumor;

                    if (rumor.content === "stop") {
                        return true
                    }
                    
                    await client.sendPrivateMsg(sender, "Echo: " + rumor.content);
                } catch (error) {
                    console.log("Impossible to decrypt DM:", error);
                }
            }
        },
        // Handle relay message
        handleMsg: async (relayUrl, message) => {
            //console.log("Received message from", relayUrl, message.asJson());
        }
    };

    client.handleNotifications(handle);
}

main();

File: sdk/examples/bot.rs
// Copyright (c) 2022-2023 Yuki Kishimoto
// Copyright (c) 2023-2025 Rust Nostr Developers
// Distributed under the MIT software license

use nostr_gossip_memory::prelude::*;
use nostr_sdk::prelude::*;

#[tokio::main]
async fn main() -> Result<()> {
    tracing_subscriber::fmt::init();

    let keys = Keys::parse("nsec12kcgs78l06p30jz7z7h3n2x2cy99nw2z6zspjdp7qc206887mwvs95lnkx")?;
    let gossip = NostrGossipMemory::unbounded();
    let client = Client::builder()
        .signer(keys.clone())
        .gossip(gossip)
        .build();

    println!("Bot public key: {}", keys.public_key().to_bech32()?);

    client.add_relay("wss://nostr.oxtr.dev").await?;
    client.add_relay("wss://relay.damus.io").await?;
    client.add_relay("wss://nostr.mom").await?;
    client.add_relay("wss://nostr.wine").await?;
    client.add_relay("wss://relay.nostr.info").await?;
    client.add_relay("wss://auth.nostr1.com").await?;

    client.connect().await;

    let subscription = Filter::new()
        .pubkey(keys.public_key())
        .kind(Kind::GiftWrap)
        .limit(0); // Limit set to 0 to get only new events! Timestamp::now() CAN'T be used for gift wrap since the timestamps are tweaked!

    client.subscribe(subscription).await?;

    client
        .handle_notifications(|notification| async {
            if let ClientNotification::Event { event, .. } = notification {
                if event.kind == Kind::GiftWrap {
                    match UnwrappedGift::from_gift_wrap(&keys, &event).await {
                        Ok(UnwrappedGift { rumor, sender }) => {
                            if rumor.kind == Kind::PrivateDirectMessage {
                                let content: String = match rumor.content.as_str() {
                                    "/hello" => String::from("Hello, World!"),
                                    "/help" => help(),
                                    _ => String::from(
                                        "Invalid command, send /help to see all commands.",
                                    ),
                                };

                                // Send private message
                                let msg =
                                    EventBuilder::private_msg(&keys, sender, content, []).await?;
                                client.send_event(&msg).to_nip17().await?;
                            }
                        }
                        Err(e) => tracing::error!("Impossible to decrypt direct message: {e}"),
                    }
                }
            }
            Ok(false) // Set to true to exit from the loop
        })
        .await?;

    Ok(())
}

fn help() -> String {
    let mut output = String::new();
    output.push_str("Commands:\n");
    output.push_str("/hello - Hello\n");
    output.push_str("/help - Help");
    output
}
